File: sso/php/sso.php
<?php
define('DISQUS_SECRET_KEY', '123456');
define('DISQUS_PUBLIC_KEY', 'abcdef');
 
$data = array(
        "id" => $user["id"],
        "username" => $user["username"],
        "email" => $user["email"]
    );
 
function dsq_hmacsha1($data, $key) {
    $blocksize=64;
    $hashfunc='sha1';
    if (strlen($key)>$blocksize)
        $key=pack('H*', $hashfunc($key));
    $key=str_pad($key,$blocksize,chr(0x00));
    $ipad=str_repeat(chr(0x36),$blocksize);
    $opad=str_repeat(chr(0x5c),$blocksize);
    $hmac = pack(
                'H*',$hashfunc(
                    ($key^$opad).pack(
                        'H*',$hashfunc(
                            ($key^$ipad).$data
                        )
                    )
                )
            );
    return bin2hex($hmac);
}
 
$message = base64_encode(json_encode($data));
$timestamp = time();
$hmac = dsq_hmacsha1($message . ' ' . $timestamp, DISQUS_SECRET_KEY);
?>
<script type="text/javascript">
var disqus_config = function() {
    this.page.remote_auth_s3 = "<?php echo "$message $hmac $timestamp"; ?>";
    this.page.api_key = "<?php echo DISQUS_PUBLIC_KEY; ?>";
}
</script>

File: sso/elixir/sso.ex
defmodule Disqus.SSO do
  @moduledoc """
  This module depends on Poison
  Add to the mix.exs deps section the folowing line
    {:poison, "~> 2.2.0"}

  like:
  defp deps do
    [
      {:poison, "~> 2.2.0"}
    ]
  end
  """


  @doc """
  user_params - a map contains:
    - required keys "id", "username", "email",
    - optional keys "avatar", "url"
  """
  def get_disqus_sso(user_params, disqus_api_key \\ nil, disqus_secret_key \\ nil) do
      # Getting keys from system env
      disqus_api_key =  disqus_api_key || System.get_env("DISQUS_API_KEY")
      disqus_secret_key = disqus_secret_key || System.get_env("DISQUS_SECRET_KEY")

      # create a JSON packet of our data attributes
      # and base64 encode
      message = %{
        "id" => user_params["id"],
        "username" => user_params["username"],
        "email" => user_params["email"],
        "avatar" => user_params["avatar"],
        "url" => user_params["url"]
      }
      |> Poison.encode!()
      |> Base.encode64()

      # generate a timestamp for signing the message
      timestamp = :os.system_time(:seconds)
      # generate our hmac signature
      hmac = :crypto.hmac(:sha, disqus_secret_key, "#{message} #{timestamp}") |> Base.encode16()

      # return a script tag to insert the sso message
      """
      <script type='text/javascript'>
          var disqus_config = function() {
              this.page.remote_auth_s3 = '#{message} #{hmac} #{timestamp}';
              this.page.api_key = '#{disqus_api_key}';
          }
  	  </script>
      """
  end
end
